### PR TITLE
Deploy es-kube-controllers in a multi-tenant environment

### PR DESCRIPTION
**Pull request by asincu**
## Description

Deploy es kube controllers in a tenant namespace to enable license copy in the managed clusters. We will use a new configuration `managedclusterlicensing`, that only deploy the licensing controller. We also need to impersonate kube-controllers service account from calico-system namespace, since this is the one that has rights inside the managed cluster.

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 1]
```diff
@@ -109,13 +111,32 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
...
 	// Watch secrets this controller cares about.
+	// ALINA : - Need to remove all esgateway watches for single tenant
 	secretsToWatch := []string{
 		render.TigeraElasticsearchGatewaySecret,
 		monitor.PrometheusClientTLSSecretName,
 	}
-	for _, ns := range []string{common.OperatorNamespace(), render.ElasticsearchNamespace} {
+
+	// Determine namespaces to watch.
+	namespacesToWatch := []string{esGatewayHelper.TruthNamespace(), esGatewayHelper.InstallNamespace()}
```
Comment (caseydavenport) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:134]: There's a utility function for this - `tenancy.GetWatchNamespaces`

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 2]
```diff
@@ -149,22 +176,40 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
...
+		return reconcile.Result{}, nil
+	}
+
+	// When running in multi-tenant mode, we need to install es-kubecontrollers in tenant Namespaces. However, the LogStorage
+	// resource is still cluster-scoped (since ES is a cluster-wide resource), so we need to look elsewhere to determine
+	// which tenant namespaces require an es-kubecontrollers instance. We use the tenant API to determine the set of
+	// namespaces that should have an es-kubecontrollers.
+	tenant, _, err := utils.GetTenant(ctx, r.multiTenant, r.client, request.Namespace)
+	if errors.IsNotFound(err) {
+		reqLogger.Info("No Tenant in this Namespace, skip")
```
Comment (caseydavenport) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:202]: This should be `.V(1)` I think to avoid being too spammy.

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 3]
```diff
@@ -280,20 +326,22 @@ func (r *ESKubeControllersController) Reconcile(ctx context.Context, request rec
...
-		pullSecrets,
-		hdler,
-		reqLogger,
-		trustedBundle,
-		r.usePSP,
-	); err != nil {
-		return reconcile.Result{}, err
+	if !r.multiTenant {
+		// ESGateway is required in order for kube-controllers to operate successfully, since es-kube-controllers talks to ES
+		// via this gateway. However, in multi-tenant mode we disable es and so this isn't needed.
```
Comment (caseydavenport) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:331]: >  in multi-tenant mode we disable es and so this isn't needed.

probably

> in multi-tenant mode, es-kube-controllers doesn't talk to elasticsearch, so this is only needed in single-tenant clusters.

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 4]
```diff
@@ -313,6 +361,7 @@ func (r *ESKubeControllersController) Reconcile(ctx context.Context, request rec
 		TrustedBundle:                trustedBundle,
 		Namespace:                    helper.InstallNamespace(),
 		BindingNamespaces:            namespaces,
+		MultiTenant:                  r.multiTenant,
```
Comment (caseydavenport) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:364]: We should be passing `Tenant: tenant` here instead, so we can use `tenant.MultiTenant()`

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 5]
```diff
@@ -111,13 +116,27 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
...
+	if opts.MultiTenant {
+		if err = c.Watch(&source.Kind{Type: &operatorv1.Tenant{}}, &handler.EnqueueRequestForObject{}); err != nil {
+			return fmt.Errorf("log-storage-kubecontrollers failed to watch Tenant resource: %w", err)
+		}
+	}
+
+	// The namespace(s) we need to monitor depend upon what tenancy mode we're running in.
+	// For single-tenant, everything is installed in the calico-system namespace.
+	// Make a helper for determining which namespaces to use based on tenancy mode.
+	esGatewayHelper := utils.NewNamespaceHelper(opts.MultiTenant, render.ElasticsearchNamespace, "")
```
Comment (rene-dekker) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:128]: Wouldn't it be better to do:
```
	esgwNamespace := utils.NewNamespaceHelper(opts.MultiTenant, render.ElasticsearchNamespace, "").InstallNamespace()
	kubecontrollersNamespace := utils.NewNamespaceHelper(opts.MultiTenant, common.CalicoNamespace, "").InstallNamespace()

```

And move the lines to right above the place they are used? Imo the code is good enough to do without the comment, but I don't mind keeping it either.

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 6]
```diff
@@ -151,22 +176,40 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
...
 }
 
 func (r *ESKubeControllersController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	helper := utils.NewNamespaceHelper(false, common.CalicoNamespace, request.Namespace)
+	helper := utils.NewNamespaceHelper(r.multiTenant, common.CalicoNamespace, request.Namespace)
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name, "installNS", helper.InstallNamespace(), "truthNS", helper.TruthNamespace())
 	reqLogger.Info("Reconciling LogStorage - ESKubeControllers")
 
+	// We skip requests without a namespace specified in multi-tenant setups.
+	if r.multiTenant && request.Namespace == "" {
```
Comment (rene-dekker) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:192]: Does that mean we don't need to watch objects without a namespace either?

File: pkg/render/kubecontrollers/kube-controllers.go [hunk 7]
```diff
@@ -171,10 +198,22 @@ func NewElasticsearchKubeControllers(cfg *KubeControllersConfiguration) *kubeCon
...
-	enabledControllers := []string{"authorization", "elasticsearchconfiguration"}
+	var enabledControllers []string
+
+	if !cfg.Tenant.MultiTenant() {
+		enabledControllers = append(enabledControllers, "authorization", "elasticsearchconfiguration")
+		if cfg.ManagementCluster != nil {
+			enabledControllers = append(enabledControllers, "managedcluster")
+		}
+	} else {
+		if cfg.ManagementCluster != nil {
```
Comment (rene-dekker) [inline pkg/render/kubecontrollers/kube-controllers.go:209]: Could be else-if to simplify the code. But if cfg.Tenant.MultiTenant()=True, doesn't that always mean that cfg.ManagementCluster != nil ?

File: pkg/render/kubecontrollers/kube-controllers.go [hunk 8]
```diff
@@ -578,7 +625,10 @@ func (c *kubeControllersComponent) controllersDeployment() *appsv1.Deployment {
 			},
 		},
 	}
-	render.SetClusterCriticalPod(&(d.Spec.Template))
+
+	if !c.cfg.Tenant.MultiTenant() {
+		render.SetClusterCriticalPod(&(d.Spec.Template))
```
Comment (rene-dekker) [inline pkg/render/kubecontrollers/kube-controllers.go:630]: I believe these are unnecessary parentheses.

**Review by rene-dekker — APPROVED** [review]
lgtm

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go (hunk 6 above)
Comment (caseydavenport) [reply pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:198]: > Does that mean we don't need to watch objects without a namespace either in multi-tenant setups?

We need to watch them, but with the `EnqueueAllTenants` handler instead.

This particular check is probably superfluous to be honest - if all is working as intended, we should never hit this branch. However, I think it's a fine-enough safety valve.

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 9]
```diff
@@ -272,40 +322,43 @@ func (r *ESKubeControllersController) Reconcile(ctx context.Context, request rec
...
-		install,
-		variant,
-		pullSecrets,
-		hdler,
-		reqLogger,
-		gwTrustedBundle,
-		r.usePSP,
-	); err != nil {
-		return reconcile.Result{}, err
+	if !r.multiTenant {
```
Comment (caseydavenport) [inline pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:331]: This needs the same comment:

> 	// In zero-tenant clusters, ESGateway is required in order for kube-controllers to operate successfully, since es-kube-controllers talks to ES via this gateway. However, in multi-tenant mode we disable the elasticsearch controller and so this isn't needed.

Comment (caseydavenport) [reply pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:331]: I see that you moved it below actually, but I think it's better at the top of the branch. It should also say "In zero-tenant clusters" for clarity.

File: pkg/render/kubecontrollers/kube-controllers.go [hunk 10]
```diff
@@ -100,6 +103,9 @@ type KubeControllersConfiguration struct {
 
 	// List of namespaces that are running a kube-controllers instance that need a cluster role binding.
 	BindingNamespaces []string
+
+	// Whether or not to run the rendered components in multi-tenant mode.
```
Comment (caseydavenport) [inline pkg/render/kubecontrollers/kube-controllers.go:107]: This comment isn't correct - the Tenant object provides tenant configuration for both single and multi-tenant modes. If this is nil, then we should run in zero-tenant mode.

File: pkg/render/kubecontrollers/kube-controllers.go [hunk 11]
```diff
@@ -171,10 +198,20 @@ func NewElasticsearchKubeControllers(cfg *KubeControllersConfiguration) *kubeCon
...
+		}
+	} else {
+		enabledControllers = append(enabledControllers, "managedclusterlicensing")
+	}
 
-	if cfg.ManagementCluster != nil {
-		enabledControllers = append(enabledControllers, "managedcluster")
+	kubeControllerConfigName := "elasticsearch"
+	if cfg.Tenant.MultiTenant() {
+		kubeControllerConfigName = fmt.Sprintf("tenant-%s", cfg.Tenant.Spec.ID)
```
Comment (caseydavenport) [inline pkg/render/kubecontrollers/kube-controllers.go:214]: I think this code should be reverted - we're not creating one per tenant any more IIUC

File: pkg/render/kubecontrollers/kube-controllers.go [hunk 12]
```diff
@@ -578,7 +623,10 @@ func (c *kubeControllersComponent) controllersDeployment() *appsv1.Deployment {
 			},
 		},
 	}
-	render.SetClusterCriticalPod(&(d.Spec.Template))
+
+	if !c.cfg.Tenant.MultiTenant() {
```
Comment (caseydavenport) [inline pkg/render/kubecontrollers/kube-controllers.go:627]: Is this change needed? Why don't we want to set this in multi-tenant?

File: pkg/render/kubecontrollers/kube-controllers.go (hunk 11 above)
Comment (asincu) [reply pkg/render/kubecontrollers/kube-controllers.go:214]: Yup. I have this removed, forgot to push.

File: pkg/render/kubecontrollers/kube-controllers.go (hunk 12 above)
Comment (asincu) [reply pkg/render/kubecontrollers/kube-controllers.go:627]: I cannot create multiple instances of es-kubecontrollers with this setting.

Comment (caseydavenport) [reply pkg/render/kubecontrollers/kube-controllers.go:624]: Okie doke.

File: pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go [hunk 13]
```diff
@@ -247,15 +293,19 @@ func (r *ESKubeControllersController) Reconcile(ctx context.Context, request rec
 	}
 
 	// Get secrets needed for kube-controllers to talk to elastic.
-	kubeControllersUserSecret, err := utils.GetSecret(ctx, r.client, kubecontrollers.ElasticsearchKubeControllersUserSecret, helper.TruthNamespace())
-	if err != nil {
-		r.status.SetDegraded(operatorv1.ResourceReadError, "Failed to get kube controllers gateway secret", err, reqLogger)
-		return reconcile.Result{}, err
+	var kubeControllersUserSecret *core.Secret
```
[the remark this answers is not in the document]
Comment (caseydavenport) [reply pkg/controller/logstorage/kubecontrollers/es_kube_controllers.go:293]: Typo: "eskube controllers" should be "es-kube-controllers"

**Review by caseydavenport — APPROVED** [review]
@asincu one typo but otherwise LGTM
